core/jvm/src/main/scala/fs2/compression.scala

```diff
@@ -10,9 +10,175 @@ import cats.effect.Sync
 /** Provides utilities for compressing/decompressing byte streams. */
 object compression {
 
+  object ZLibParams {
+
+    sealed abstract class Header(private[compression] val juzDeflaterNoWrap: Boolean)
+    case object Header {
+      private[compression] def apply(juzDeflaterNoWrap: Boolean): ZLibParams.Header =
+        if (juzDeflaterNoWrap) ZLibParams.Header.GZIP else ZLibParams.Header.ZLIB
+
+      final case object ZLIB extends Header(juzDeflaterNoWrap = false)
+      final case object GZIP extends Header(juzDeflaterNoWrap = true)
+    }
+
+  }
+
+  /**
+    * Deflate algorithm parameters for the selected implementation, as one of:
+    *   1. [[compression.JavaUtilZipDeflaterParams]] for [[java.util.zip.Deflater]]
+    */
+  sealed trait DeflateParams
+
+  /**
+    * Deflate algorithm parameters for the [[java.util.zip.Deflater]] implementation.
+    */
+  sealed trait JavaUtilZipDeflaterParams extends DeflateParams {
```

Comment:
Is this `JavaUtilZipDeflaterParams` intended to account for a possible future where `deflate` is implemented without using `java.util.zip.Deflate` and hence may require different config? If so, I think I'd prefer to skip that and just merge this in to `DeflateParams`. Similar on inflate side.

Looks good overall though, what do you think?

Reply:
Yes, I have looked at alternative implementations and found that they have different configuration. As one example, the java.util.zip implementation only supports some of the flush modes requested in #1831, in fact there are several things it doesn't support or supports differently (eg. compression levels), so it may not be possible to support alternative implementations with a single parameter set. The intention of the naming of these two Params classes is for the provided configuration to also determine the desired implementation, where one of multiple may be selected.

As I understand it, you would prefer that this not support multiple future implementations, but only one, where if it is changed it will need to support at least the same functionality? No problem, I'll reduce and refactor as suggested, and update the PR.

As an aside and explanation, and depending on the time I'll have to put into these late night experiments and musings, my gut feeling is that there may be a performance gain to be made here by replacing java.util.zip's buffering-, thread-safe and JNI-based implementation with a pure fs2-based implementation. This would work by directly streaming the deflated/inflated Huffman encoding/decodings without first copying and concatenating these into the fixed-size output buffer controlled by the bufferSize parameter, also avoiding synchronization and native calls. If this is possible it should increase performance by avoiding buffer-filling array copying and may have interesting memory characteristics if the original and immutable Huffman encoding/decoding elements are directly streamed.

Reply:
@mpilquist - PR updated as suggested with simplified DeflateParams and InflateParams. Anything else? Thanks.